### PR TITLE
Enhancement Proposal EP02: Add Automated Rendering to fc4-tool

### PR DESCRIPTION
**Pull request by aviflax**
Please see #72 for context.

### CODE REVIEW

**Comment by AndreaCrotti** [comment]
I think this would be very important, but I suspect that using FC4 any kind of automation would be a kind of a hack, since it's not an open source/free tool after all.

One notable missing thing from the examples ther e is graphviz/dot: 
https://graphviz.gitlab.io/gallery/

which is available everywhere, it's OSS and has lots of libraries for every language.
The only "challenge" would be maybe to define some HTML styles for the various components which can then be shared/abstracted, but once that is done it might be a serious contender for complex graphs imho.

The only advantage I would see in some more Web driven solutions imho is that we could potentially do dynamic graphs (like the ones done with https://d3js.org/) which would allow to play around, zoom in and out of complex graphs for example, but FC4 doesn't really help with that I think?

File: proposals/ep02-automated-rendering/ep02-automated-rendering.md [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+[Batik](https://xmlgraphics.apache.org/batik/) is:
+
+> a Java-based toolkit for applications or applets that want to use images in the Scalable Vector Graphics (SVG) format for various purposes, such as display, generation or manipulation.
+
+…so I’d guess we’d be able to use it to generate “native” vector graphics and export them as PDF and PNG. The most compelling advantage here would be the ability to render images without any external dependencies such as Node, Puppeteer, Chrome, etc.
+
+### Quil
+
+[Quil](http://quil.info/) is a Clojure/ClojureScript library for creating interactive drawings and animations. It’s based on [Processing](https://processing.org/). I don’t remember exactly why, but at some point I thought it might be worth trying to use it to implement diagram rendering “from scratch” i.e. at a very low level. I have an old branch somewhere with a single commit in it, but really there’s barely anything in there. It’s really just a stub of an idea.
```
[the remark this answers is not in the document]
Comment (aviflax) [reply proposals/ep02-automated-rendering/ep02-automated-rendering.md]: Makes sense to me. Thanks!

**Comment by aviflax** [comment]
@AndreaCrotti 

> I think this would be very important

OK, that’s really helpful to know, thanks!

> but I suspect that using FC4 any kind of automation would be a kind of a hack, since it's not an open source/free tool after all.

Assuming you mean Structurizr Express, I hear you. Even though its creator suggested it, using browser scripting to automate a Web GUI could be quite brittle. Any Web GUI, especially a free, unsupported one, is subject to refactoring and/or decommissioning, in which case our automation would break.

OTOH, since Structurizr Express is an SPA, and it runs entirely client-side, it’d probably be viable to “vendor” it in its entirety, in a working state, into fc4-tool. I mean, it’s basically just a bunch of HTML, CSS, and JavaScript files.

I’m not saying that using Structurizr Express is especially promising or attractive, but I do still consider it a viable option (pending a PoC of course).

(As for the full-blown Structurizr product, that’s commercial software with supported APIs, so _if_ we were willing to pay for it and make rendering dependent on a working Internet connection, then I think that might be a viable option as well.)

> One notable missing thing from the examples there is graphviz/dot…
 
I’m well aware of Graphviz. I’ve tried to use it for software architecture diagramming _many_ times over the years, and I’ve seen numerous results of other people’s attempts, and I’ve never been satisfied by the results.

I discussed this in my recent blog post [The FC4 Framework](https://engineering.fundingcircle.com/blog/2018/09/07/the-fc4-framework/):

> The few tools that do have a textual representation and a command-driven rendering process — the most prominent being Graphviz — have never felt like a good fit for diagramming software architectures. Graphviz, for example, being focused on graphs, makes composition and nesting quite awkward, and supports only algorithmic layout when rendering diagrams, making it very difficult to tell a coherent story with the relative positioning of elements. Whenever I tried Graphviz, I’d find myself wrestling with its algorithms just to get things laid out clearly and usefully.

----

> The only advantage I would see in some more Web driven solutions imho is that we could potentially do dynamic graphs (like the ones done with https://d3js.org/) which would allow to play around, zoom in and out of complex graphs for example, but FC4 doesn't really help with that I think?

Well, FC4 currently does [currently support](https://fundingcircle.github.io/fc4-framework/methodology/scheme.html) Structurizr’s [Dynamic diagrams](https://structurizr.com/help/dynamic-diagrams), which are really just animated workflow diagrams. But that support is mostly due to how FC4 is currently so dependent on Structurizr Express, and so tightly coupled to it, that it was mostly free to declare that type of diagram supported, and so I included it in the FC4 methodology because some people at Funding Circle wanted to use it. Their usefulness within FC4 is quite limited, however, because the artifacts currently produced via FC4 are exclusively static PNG files; the animation is only viewable from within Structurizr Express.

Sorry, that was TMI. Bottom line: I think that for the near future FC4 will remain focused on producing static images (PNG and/or PDF); IMHO neither animation nor interactivity should be considered anytime soon. That said, I’d love to hear your thoughts on this, and on all of the above — what am I missing? What do you agree or disagree with?

Thanks!

**Comment by AndreaCrotti** [comment]
I'm perfectly fine with PNGs, I was just saying that the only advantage of these more "fancy" tools like Fc4 over things like Graphviz are imho the potential capability to do more dynamic graphs.

And well I understand your point about Graphviz and not producing the nicest of results, but on the other hand it's a very well established tool and quite powerful.

**Comment by AndreaCrotti** [comment]
One more general doubt I have is what are these graphs for.
I mean if you create a graph from code you don't get too much saying about the layout that will be chosen to render it (unless you use something like SVG I guess).
These are great and easy to produce (specially with graphviz there is a lot of tooling already there), but in some cases like our kafka streams apps, maybe we do want to present things in a certain order.

So everyone just ends up using LucidChart for this kind of stuff, which sucks for many reasons but at least it allows you to have full control.
Something like FC4 seems like a middle ground, because you lose the control of how things are rendered but you don't get the automation/tooling you have with graphviz & co.

I don't really have a solution unfortunately, it looks like it's a hard/unsolvable problem, and maybe everyone will continue using different tools depending on what they need.

**Comment by aviflax** [comment]
> I'm perfectly fine with PNGs…

Cool.

> I was just saying that the only advantage of these more "fancy" tools like Fc4 over things like Graphviz are imho the potential capability to do more dynamic graphs.

Did you mean Structurizr Express when you wrote FC4 here? Because I’m confused. Structurizr Express is a diagramming tool, somewhat equivalent to Graphviz, but I consider FC4 something quite different; it’s a framework that couples a methodology _and_ a tool. (Right now that tool happens to be Structurizr Express, but I’m hoping to reduce its usage in the framework and eventually remove it from the framework.)

Anyway, I agree that interactivity and animation are indeed advantages that some of these rendering options have over Graphviz, but I don’t agree that they’re the only advantage. I’ll elaborate below.

> And well I understand your point about Graphviz and not producing the nicest of results, but on the other hand it's a very well established tool and quite powerful.

It certainly is. I just think that Graphviz and FC4 are designed to solve different problems. Graphviz is focused on algorithmic visualizations of _mathematical_ graphs — sets of nodes and edges. But FC4, the C4 Model, and Structurizr Express are all focused on _crafted_ visualizations — diagrams — of software systems and their relationships. For these free-form absolute positioning is critical and algorithmic layout is unhelpful.

I’m not saying it wouldn’t be _possible_ to render useful software architecture diagrams with Graphviz. I’m saying I think doing so goes against the “grain” of the tool, and is therefore onerous and frustrating, and is unlikely to produce the highest-quality end results.

> One more general doubt I have is what are these graphs for.

I think this is the crux of the matter. As I see it, FC4 only _incidentally_ involves graphs — the relationships between our users, software systems, and containers. Those relationships are crucial and need to be included in the diagrams, but ultimately an FC4 diagram is a _diagram_ and _not_ a graph.

> So everyone just ends up using LucidChart for this kind of stuff, which sucks for many reasons but at least it allows you to have full control.

Exactly! This is the problem I’m trying to solve with FC4 — we need free-form absolute positioning and encapsulation in order to produce effective software architecture diagrams, but there are all sorts of problems with making these diagrams with a GUI — problems I’ve described in depth in [my blog post](https://engineering.fundingcircle.com/blog/2018/09/07/the-fc4-framework/).

> Something like FC4 seems like a middle ground…

That’s certainly what I have in mind — I want it to (eventually) have the best properties of free-form GUI apps like Lucidchart and text-driven rendering systems like Graphviz and SVG, but custom-tailored _specifically_ for software architecture diagramming (hopefully soon with [its own DSL](https://github.com/FundingCircle/fc4-framework/pull/74)) and hopefully without some of the biggest downsides of those other solutions.

> …because you lose the control of how things are rendered…

I’m not sure I know what you mean here? I do feel like FC4 even in its current state wherein it uses  Structurizr Express for rendering already provides a lot of flexibility as to layout and styling.

> …but you don't get the automation/tooling you have with graphviz & co.

Not _yet_ you don’t. That’s kinda the whole point of this proposal — to implement automated rendering!

**Comment by AndreaCrotti** [comment]
Yes sorry I was talking about Structurizr Express not FC4.
And ok yes I had originally misunderstood how this tool was going to be used and that's why I didn't even see you could position elements where you wanted them.

Anyway I mentioned Graphviz just because you also mentioned other alternatives and I have the feeling that any possible way to automate structurizr express would be kind of a hack, since it's a closed source really.

**Comment by aviflax** [comment]
While I’m waiting for feedback on this EP and #72 and the other EPs in this batch, I’ve been doing some PoCs of a few of the approaches described in this EP. I tried using Paper.js in Node with jsdom and got as far as rendering a few basic shapes, but then I learned that text rendering would be a major problem as Paper.js has no support for wrapping text. I tried improving the JGraphx branch but hit some bugs in its text rendering as well. So then I decided to try rendering with Structurizr Express in a headless browser with Puppeteer, and I’ve had some really promising results there. In fact I’ve more or less got it basically working in [this branch](https://github.com/FundingCircle/fc4-framework/tree/automated-rendering).

So I think I’m going to revise this EP to make that approach the recommended approach. In the revision I’ll explain why, of course. Then I’ll focus on putting the finishing touches on the implementation — which should be mainly updating the docs and the CI config, as I’ve already got an automated test that’s passing. (Oh all right, I’ll add a sad path test as well, and maybe some error handling too ;)

**Comment by simonbrowndotje** [comment]
> While I’m waiting for feedback on this EP and #72 and the other EPs in this batch, I’ve been doing some PoCs of a few of the approaches described in this EP. I tried using Paper.js in Node with jsdom and got as far as rendering a few basic shapes, but then I learned that text rendering would be a major problem as Paper.js has no support for wrapping text. I tried improving the JGraphx branch but hit some bugs in its text rendering as well. So then I decided to try rendering with Structurizr Express in a headless browser with Puppeteer, and I’ve had some really promising results there. In fact I’ve more or less got it basically working in [this branch](https://github.com/FundingCircle/fc4-framework/tree/automated-rendering).

Yes, as you've discovered, rendering diagrams in the browser with HTML/CSS/JavaScript is not particularly straightforward, especially across browsers. Exporting those diagrams to PNG is trickier still. Structurizr uses JointJS, but also take a look at Mermaid and jsPlumb. Alternatively, if you can live with automatic layouting, you can get some decent diagrams from PlantUML. The Structurizr for Java/.NET libraries already have some exporters, that you could use as a starting point -> https://github.com/structurizr/java/blob/master/structurizr-plantuml/src/com/structurizr/io/plantuml/PlantUMLWriter.java

**Comment by simonbrowndotje** [comment]
> The only advantage I would see in some more Web driven solutions imho is that we could potentially do dynamic graphs (like the ones done with https://d3js.org/) which would allow to play around, zoom in and out of complex graphs for example, but FC4 doesn't really help with that I think?

Yes @AndreaCrotti ... there's a lot that you do with d3, especially once you start creating models of your software systems, as opposed to a collection of individual diagrams (i.e. separating models from views). Structurizr (the API-driven, full version) also includes a number of d3-based visualisations that you can use, even on the free plan -> https://structurizr.com/help/explorations ... and examples here -> https://structurizr.com/share/1/explore/graph (you can change the visualisation via the dropdown)

**Comment by aviflax** [comment]
> Yes, you could use something like headless Chrome and Puppeteer to automate some aspects of Structurizr Express ... but I certainly don't recommend this.

I’ve actually been giving this a try and it’s actually looking fairly promising… I’d like to discuss it with you via email if that’s OK… I’ll shoot you a note on Monday.

> Yes, as you've discovered, rendering diagrams in the browser with HTML/CSS/JavaScript is not particularly straightforward, especially across browsers.

For sure!

> Exporting those diagrams to PNG is trickier still. Structurizr uses JointJS, but also take a look at Mermaid and jsPlumb. Alternatively, if you can live with automatic layouting, you can get some decent diagrams from PlantUML. The Structurizr for Java/.NET libraries already have some exporters, that you could use as a starting point

Thank you for all the help!

> …there's a lot that you [can] do with d3, especially once you start creating models of your software systems, as opposed to a collection of individual diagrams (i.e. separating models from views).

This makes sense, and I’ve come around on this BTW, as per #73.

> Structurizr (the API-driven, full version) also includes a number of d3-based visualisations that you can use, even on the free plan…

😎 very cool, I’m looking forward to checking them out!

**Comment by aviflax** [comment]
As per my changes to this EP I believe it has been successfully implemented. At this point I’m just looking for one person to agree with me, review the EP text for any glaring errors, and hopefully approve this PR so I can merge it. @creese @99-not-out @kidpollo perhaps one of you could take a look? Thanks!

File: proposals/README.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# Enhancement proposals
+
+This directory tree hosts enhancement proposals.
```
Comment (sgerrand) [inline proposals/README.md]: 🙋 This branch needs to be rebased on `master`.

File: proposals/ep02-automated-rendering/ep02-automated-rendering.md [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+  </tr>
+</table>
+
+## Summary
+
+The [current workflow](https://fundingcircle.github.io/fc4-framework/methodology/authoring_workflow.html) for rendering diagram images from their YAML source with fc4-tool is completely manual, onerous, tedious, error-prone, hard to learn, and inefficient. We should replace it with a new workflow enabled by automated image rendering.
+
+## Status
+
+The [recommended approach](#recommended-implementation-approach) was implemented in mid-November 2018. Key pull requests are labeled with [EP02](https://github.com/FundingCircle/fc4-framework/pulls?utf8=%E2%9C%93&q=is%3Apr+label%3AEP02+).
```
Comment (sgerrand) [inline proposals/ep02-automated-rendering/ep02-automated-rendering.md]: 💭 An errant `+` appears at the end of https://github.com/FundingCircle/fc4-framework/pulls?utf8=%E2%9C%93&q=is%3Apr+label%3AEP02

File: proposals/ep02-automated-rendering/ep02-automated-rendering.md [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+
+
+## New User Experience (Workflow)
+
+1. The documentarian runs a command such as:
```
Comment (sgerrand) [inline proposals/ep02-automated-rendering/ep02-automated-rendering.md]: 🙋 I think you should define this term or link to the version you are referencing. The OE and MW dictionaries [define it differently](https://www.merriam-webster.com/dictionary/documentarian).

File: proposals/ep02-automated-rendering/ep02-automated-rendering.md [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+   1. Every specified view would be rendered to a PNG file in the same directory and with the same file name as the YAML file, except for the extension.
+
+## Desired Implementation Properties
+
+I’m not sure if these are all possible, so let’s think of this as a wish list for now:
+
+* A: Operates completely locally; does not require any sort of network connectivity
+* B: Has as few additional system requirements as possible, ideally zero
+* C: Operates completely in a single OS process
+* D: Requires adding no additional programming systems to the tool
```
Comment (sgerrand) [inline proposals/ep02-automated-rendering/ep02-automated-rendering.md]: 💭 Is this an unordered _and_ an ordered list at once?

File: proposals/ep02-automated-rendering/ep02-automated-rendering.md [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+  		<td>✓</td>
+  		<td>✓</td>
+      <td>✓</td>
+  	</tr>
+  </tbody>
+</table>
+
+## On ClojureScript
+
+fc4-tool is currently implemented with “just Clojure” i.e. the O.G. variant of [Clojure](https://clojure.org) that runs on the JVM. That’s as opposed to the other popular variant of Clojure, [ClojureScript](https://clojurescript.org), which is transpiled to JavaScript and thereby runs in browsers and/or [Node.js](https://nodejs.org/). The main reason for this is simply that it’s what I knew, so it was easiest for me. Ease matters sometimes, to some degree, but there are many instances wherein it is outweighed by other priorities, and this might be one of them. In the event that some approach for implementing automated rendering is highly compelling but would require running in Node, I’d be open to considering migrating all of fc4-tool to ClojureScript. (This might yield additional benefits such as reduced execution time.)
```
Comment (sgerrand) [inline proposals/ep02-automated-rendering/ep02-automated-rendering.md]: 🙋 I'd replace "O.G." with "original", dawg.

File: proposals/ep02-automated-rendering/ep02-automated-rendering.md (hunk 3 above)
Comment (aviflax) [reply proposals/ep02-automated-rendering/ep02-automated-rendering.md]: Thanks! Will fix.

File: proposals/ep02-automated-rendering/ep02-automated-rendering.md (hunk 5 above)
Comment (aviflax) [reply proposals/ep02-automated-rendering/ep02-automated-rendering.md]: LOL, I suppose it is. 😜

File: proposals/ep02-automated-rendering/ep02-automated-rendering.md (hunk 4 above)
Comment (aviflax) [reply proposals/ep02-automated-rendering/ep02-automated-rendering.md]: Good point! Thanks!

File: proposals/README.md (hunk 2 above)
Comment (aviflax) [reply proposals/README.md]: Will do, thanks!

File: proposals/ep02-automated-rendering/ep02-automated-rendering.md (hunk 6 above)
Comment (aviflax) [reply proposals/ep02-automated-rendering/ep02-automated-rendering.md]: 😜

File: proposals/ep02-automated-rendering/ep02-automated-rendering.md (hunk 4 above)
Comment (aviflax) [reply proposals/ep02-automated-rendering/ep02-automated-rendering.md]: Added [this link](http://www.writethedocs.org/documentarians/) in 6a1bcba — thanks!

File: proposals/ep02-automated-rendering/ep02-automated-rendering.md (hunk 6 above)
Comment (aviflax) [reply proposals/ep02-automated-rendering/ep02-automated-rendering.md]: Fixed in 6a1bcba

File: proposals/ep02-automated-rendering/ep02-automated-rendering.md (hunk 3 above)
Comment (aviflax) [reply proposals/ep02-automated-rendering/ep02-automated-rendering.md]: Fixed in 6a1bcba — thanks!

**Pull request merged**